### PR TITLE
[AIRFLOW-3611] Breeze

### PR DESCRIPTION
**Pull request by potiuk**
~This is a Draft pull request for initial comments. It's not yet intended to be merged.~

The intention is to merge it :)

### CODE REVIEW

**Comment by potiuk** [comment]
Comment to self (@mik-laj ) - convert the doc to .rst format. @mik-laj  - maybe you will help with that ?? I am happy to get your help on that one.

File: setup.py
Comment (andriisoldatenko) [inline setup.py]: I think better to show, that it's python version dependent:
```
if not PY3:
    FileNotFoundError = IOError
```

File: setup.py
Comment (andriisoldatenko) [inline setup.py]: how do you think to remove `Custom clean` and rephrase:
`Command to tidy up the project root.`

Comment (andriisoldatenko) [reply setup.py]:
```
In [6]: help(CleanCommand)
Help on class CleanCommand in module __main__:

class CleanCommand(Command)
 |  Command to tidy up the project root.
 |
 |  Method resolution order:
```

File: .hadolint.yaml
Comment (ashb) [inline .hadolint.yaml]: Hadolint is GPL'd, so we'll need to obtain an exemption from the Apache Legal team to use this.

File: Dockerfile
Comment (ashb) [inline Dockerfile]: It's possible to set this directly via env variables https://pip.pypa.io/en/stable/user_guide/#environment-variables

```suggestion
ENV PIP_NO_CACHE_DIR=${PIP_NO_CACHE_DIR}
```

and then we don't need to add it in to the `pip` invocations

File: Dockerfile
Comment (ashb) [inline Dockerfile]: I know exactly what this is doing but people less familar with Docker might see `FORCE_REINSTALL_ALL_DEPENDENCIES="1"` and think `1` = enable this feature.

How about?

```suggestion
ARG DEPENDENCIES_EPOCH_NUMBER="1"
```

I've chose "epoch" here after the concept from debian version number (2:1.0-1, `2:` is the epoch)

File: Dockerfile
Comment (ashb) [inline Dockerfile]: Minor point, but probably worth installing pip before copying any of the airflow setup.py etc. files?

File: Dockerfile
Comment (ashb) [inline Dockerfile]: Why do we install mysql libs in this image but not postgres?

Also why do we need settings for mysql server when installing client packages?

File: scripts/ci/in_container/run_ci.sh
Comment (ashb) [inline scripts/ci/in_container/run_ci.sh]: Can we replace this script by setting some https://hub.docker.com/_/mysql#environment-variables (say MYSQL_DATABASE) in the docker-compose file?

**Review by ashb — COMMENTED** [review]
Non-complete look, just a few spot comments.

File: setup.py
Comment (potiuk) [reply setup.py]: Good point. Will fix.

File: setup.py
Comment (potiuk) [reply setup.py]: Fine for me. I just did not want to change this too much, but Now I am going to split this a bit more into several small PRs so maybe such fixes make more sense :)

File: .hadolint.yaml
Comment (potiuk) [reply .hadolint.yaml]: Ok. I believe that should be possible. I'd assume we are using the binary and we do not modify sources. It's no different than when we use linux kernel for example. Is there any paper-work needed to ask for exemption? I will gladly follow that as hadolint looks rather nice.

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: In this case actually we have to use the directive, because it can be more complex than --no-cache. When you build it during the Travis CI/DockerHub/Local Breeze build it will be set to use --cache-from=<DOWNLOADED_IMAGE> to use downloaded cache from DockerHub. It will only be set --no-cache when you choose to build it without cache (either via command line, setting the right environment variable before build or when you run CRON build on Travis).

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: I like it :). Will change.

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: Ah No.. It's pip cache :).... You are right... will be better this way.

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: One more layer but worth it :)

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: Not sure. I just wanted to have an exact copy of what we have in the current Dockerfile (https://github.com/apache/airflow/blob/master/Dockerfile). I can try to remove that and see what happens, but I wanted to be as close as possible. I am sure it was put there for a reason. Maybe @fokko or @ffinfo can shed some light on why the libraries were added in the original Dockerfile?

File: scripts/ci/in_container/run_ci.sh
Comment (potiuk) [reply scripts/ci/in_container/run_ci.sh]: I think this way is quite good. As we need to pass only one variable BACKEND through bash/docker-compose/docker chain. 

We could potentially set the database variables outside -  the easiest way I know is to have extra docker-compose.yaml (for example docker-compose-postgres.yaml, docker-compose-mysql.yaml and add "-f docker-compose-${BACKEND}.yaml". But I find clunky than the "if" in bash script. Especially that we use AIRFLOW_HOME for sqlite and it is defined in the Dockerfile, so it's better to set it only when we are in the container. 

Also this is abut more than database - we also choose executor (in case of SQLite) or CELERY_BACKEND.

I think this is quite good place for setting the variables.

File: setup.py
Comment (potiuk) [reply setup.py]: Fixed.

File: setup.py
Comment (potiuk) [reply setup.py]: Fixed.

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: Done.

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: Fixed.

**Comment by potiuk** [comment]
@mik-laj -> we have .rst Breeze description now.

**Comment by potiuk** [comment]
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Note to reviewer: Please review only the last commit (use commits section above). This change is based on yet unmerged #4936 change and it contains also commit from that change.  It's a github limitation unfortunately. 
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

File: .hadolint.yaml
Comment (ashb) [reply .hadolint.yaml]: https://issues.apache.org/jira/browse/LEGAL-447 was where I asked about an LGPL lib in a similar case. I don't know (haven't checked) if GPL is any different in this case.

I think my thinking last night was "GPL" is a bit more wide-reaching than LGPL so we might have to double check.

From   http://www.apache.org/legal/resolved.html#build-tools

> The most important criteria to be included in this category is that such tools are not customarily part of distributions of running code, unless such deliverables include source anyway. Other criteria for inclusion include: having been around for years, being defacto standards in their respective communities, being made available under are under "library" or "lesser" licenses or otherwise containing an exception which ensures that usage of this tool does not affect the license of the code against which it is run.

Comment (potiuk) [reply .hadolint.yaml]: Cool. I opened https://issues.apache.org/jira/browse/LEGAL-450 . If I don't hear from them soon, I might extract it out to a separate PR to not be blocked on it.

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: Done.

File: .hadolint.yaml
Comment (potiuk) [reply .hadolint.yaml]: That one was super quick: 

> As is pointed out in our FAQ: https://apache.org/legal/resolved.html#prohibited it is OK to use GPL build tools provided they leave no code traces (licensed under a cat X license) in your final release artifacts.

So I think we are safe :)

https://issues.apache.org/jira/browse/LEGAL-450?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=16796240#comment-16796240

**Comment by potiuk** [comment]
@ashb @Fokko -> I've just added a very nice and easy to setup auto-complete for breee command :). It's becoming a really useful piece of software. Also I am down to just 8 errors in Travis CI.

**Comment by potiuk** [comment]
@ashb @fokko -> as a follow up I also implemented auto-complete for "run-test" method inside the container. 

This was by far the single biggest problem I had when running the tests. Having to remember module names and the : in the right place was a major pain. Now it's super convenient to start typing the module name and use auto-complete. It goes down to individual tests.

I also renamed and shortened the run_unit_test.sh method to simple 'run-test' that you can run anywhere (it is in the path). I also added --skip-db-init flag - most of the tests do not require to re-run the database initialisation and if you want to re-run single test, it takes awfully lot of time. So you can now pass --skip-db-init flag to skip resetdb/initdb. You can even autocomplete it ;). 

I think with those changes, life of developers trying to use breeze/replicate CI tests will be much, much easier.

I used nosetests for now but we can do very similar thing with pytest when we switch to it.

File: scripts/ci/in_container/run_ci.sh
Comment (potiuk) [reply scripts/ci/in_container/run_ci.sh]: I changed it now to use separate Docker compose files. Push follows once I test it thoroughly

**Comment by potiuk** [comment]
@ashb @fokko -> I pushed latest version that is much improved:

* mysql installation is now simplified, no more server-specific configuration (thanks @ashb for the hint)
* parameters for the database are now spread across various docker-compose files. I found it easiest to have separate compose file for each database and "compose" the files when running - so for example when you run it locally you compose : docker-compose.yml + docker-compose-postgres.yml + docker-compose-local.yml (the last one adds source bind mounts for local development). 

It's nicely modular but you might not like it. so if you don't let me know (and why) - I might try to find another solution.

Also after @ashb hangout I've decided to implement a small improvement in the 'run-test' script - now the database is initialized only the first time you run test in the container, subsequent times the database is not initialized (unless you specify --with-db-init switch). This is a nice automation for the pattern you explained in your hangout, where you mentioned you rarely use run_unit_tests.sh because database initialisation is slow. With the approach I suggest, you do not have to copy&paste things but can run the tests with the scripts. Combined with auto-complete, I think it's really helpful (especially to onboard new people).

**Comment by potiuk** [comment]
Hey @ashb -> you might take a look at the latest version now. I watched your Hangout last week and it was super interesting to see how you work and how Breeze nicely automates some of the stuff that now require a lot of "wizardry" and manual setup. I took the lesson and implemented some of the stuff that you pointed out during your hangout. And the current breeze is better than ever now:

* run-test <TAB> autocomplete now also autocompletes full modules and Test Classes - so you can easily select whole module or Class rather than single tests

* run-test now only initializes the database the first time it is started in the container. When you run it subsequent times it will skip database initialisation (similar to your workflow but automated). You can always add `--with-db-init` flag to re-run initialisation if you need. 

* run-test supports now direct auto-complete-able `--nocapture` flag so that you can run tests with ipdb. You can also pass any switches to nosetest via flags following `--`
     For example `run-test tests.test_cli -- --nocapture --rednose` will work :).

* ipdb is now added as requirement, so no need to add it manually.

* I updated documentation and added chapters about debugging (both using ipdb and connecting to the IDE /IntelliJ debugger - including some screenshots and explanation). 

I hope you will find it going in the right direction. It will be much easier to onboard new people if we have all the stuff automated and do not require to know all the wizardry to make the integration tests work.

**Review by potiuk — COMMENTED** [review]
the current docs are available here: https://github.com/PolideaInternal/airflow/blob/simplified-development-workflow/BREEZE.rst

**Comment by galuszkak** [comment]
This look so cool! Great job with that @potiuk , I will try it tomorrow.

**Comment by potiuk** [comment]
Glad you like it @galuszkak -> let me know what are your experiences and I will be happy to help solving any problems you might have. I was a bit slower last few days (catching up with another project) but I hope to fix the last few remaining tests in multi-staging soon and then will try to advertise it a bit more in community :).

It would be great to have you testing it -> wondering if you use Linux. I use Mac on a daily basis and tested it a bit on Linux (in the cloud) but some "workstatation" experience on Linux would be great.

**Comment by galuszkak** [comment]
@potiuk I'm also macOS user, so I'm unable to help you with native Linux.

**Comment by potiuk** [comment]
No worries :) it will be useful anyway.

**Comment by galuszkak** [comment]
Some small feedback:
```
➜  airflow git:(simplified-development-workflow) ✗ ./breeze

 --

You are running breeze in OSX environment
The getopt version installed by OSX should be replaced by the GNU one

Run 'brew install gnu-getopt'

And link it to become default as suggested by brew by typing:
echo 'export PATH="/usr/local/opt/gnu-getopt/bin:$PATH"' >> ~/.bash_profile
. ~/.bash_profile

Login and logout afterwards


➜  airflow git:(simplified-development-workflow) ✗ export PATH="/usr/local/opt/gnu-getopt/bin:$PATH"
➜  airflow git:(simplified-development-workflow) ✗ ./breeze


===============================================================================================

         Initial diagnostics


./breeze: line 204: md5sum: command not found
Missing md5sum for /Users/galuszkak/Projects/airflow/setup.py
./breeze: line 204: md5sum: command not found
Missing md5sum for /Users/galuszkak/Projects/airflow/setup.cfg
./breeze: line 204: md5sum: command not found
Missing md5sum for /Users/galuszkak/Projects/airflow/Dockerfile
./breeze: line 204: md5sum: command not found
Missing md5sum for /Users/galuszkak/Projects/airflow/airflow/version.py
./breeze: line 204: md5sum: command not found
Missing md5sum for /Users/galuszkak/Projects/airflow/airflow/www/package.json
./breeze: line 204: md5sum: command not found
Missing md5sum for /Users/galuszkak/Projects/airflow/airflow/www/package-lock.json

!!!!!!! Your docker images are not up-to-date !!!!!!!!!!!!!!!
```

Sounds there should be also detector if someone has md5sum.

**Comment by galuszkak** [comment]
`brew install md5sha1sum`

Should be added to detect before running script.

**Comment by galuszkak** [comment]
Another error:
```

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
                     Starting step: Fixing permissions for CI builds for setup files
                     Date: Fri Apr  5 15:30:21 CEST 2019
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

/Users/galuszkak/Projects/airflow/hooks/build: line 377: gstat: command not found
/Users/galuszkak/Projects/airflow/hooks/build: line 380: gstat: command not found
```

**Comment by galuszkak** [comment]
Taking in account my error above, just install 
`brew install coreutils` on mac

**Comment by potiuk** [comment]
Thanks for those ! I will address them shortly.

**Comment by galuszkak** [comment]
To be honest, tests freeze out and stopped working at some point. So I'm not sure if this is my laptop or script.

**Comment by potiuk** [comment]
Yeah. Tests are still need to be finalized.

**Comment by potiuk** [comment]
I understand @galuszkak that you are talking about tests that you run manually? Some more information about the freeeze ? Maybe we can switch to https://apache-airflow-slack.herokuapp.com/  and chat there? I am there as "Jarek Potiuk"

**Comment by galuszkak** [comment]
@potiuk that's all from my testing. This is an amazing improvement. My first experience as a contributor, to figure out how to run locally tests was really tragic. This from contributor perspective is such a superior experience even if it's not working, it was really easy to figure it out what to do.

Thanks!

**Comment by potiuk** [comment]
@galuszkak  -> I've added detection of missing tools now :). Thanks for suggestions. BTW. It's enough to install getopt and coreutils - coreutils contains md5sum already - not only gstat.

I just provide instructions rather than run the brew or apt because you never know whether brew is installed (some people might still use port ;) ) or whether you run it on debian-derived linux.

**Comment by milton0825** [comment]
It is so much easier to run tests with breeze. It was such a pain to replicate some failing tests locally and I am glad that we have this.

Awesome work!

**Comment by potiuk** [comment]
@milton0825 -> thanks for the kind words :). I just recorded a screncast showing Breeze's basic usage ....  https://www.youtube.com/watch?v=ffKFHV6f3PQ :). I hope it will get approved by the community soon.

**Comment by galuszkak** [comment]
Hey, 

Is there any progress on this? This is a superior experience in terms of development and would be great to have it in main repo.

Thanks

**Comment by potiuk** [comment]
Yep. I continuously rebase it in on top of master with the multi-staging AIP-10. Up until recently we had a lot of instability/flaky tests in Travis  and I wanted to make sure the instability is fixed before I switch to the new CI image (which was successful eventually - we fixed a flakiness of our tests and Travis finally fixed the flakiness of Docker operations). 

Now I am working on merging the CI changes and hope to have it really soon - I just want to have confirmation from @ashb and @Fokko that the https://github.com/apache/airflow/pull/4937 is good to go and then I need to fix failing kubernetes test in https://github.com/apache/airflow/pull/4938 . 

I am travelling this week for Bay Area Meetup and a number of other meetings in Bay Area, but I hope to get it merged this week and then I have a talk about Breeze/Simplified Development Environment  https://www.meetup.com/Bay-Area-Apache-Airflow-Incubating-Meetup/ where I hope I will explain the context and benefits of Breeze and I will proceed with voting the proposal and merging it soon :).

**Comment by potiuk** [comment]
Hello @ashb @mik-laj @andriisoldatenko @dimberman @kaxil @fokko - I think breeze environment is now ready to be merged. After all the changes merged so far Breeze now becomes a swiss-army-knife of everything-testing for airflow. It builds on top of the CI images and it gives access to all related to testing, static code check (using pre-commits), image management from a single `breeze` script. I will be talking about it on Monday in NYC and doing workshops on  Tuesday so it would be great to merge it before!

**Comment by potiuk** [comment]
Comprehensive (and I think very usable) documentation here: 
https://github.com/PolideaInternal/airflow/blob/simplified-development-workflow/BREEZE.rst

Besides Breeze will guide users in many places in case of problems/missing prerequisites and the like.

**Comment by dimberman** [comment]
Wooo! Ok I'm gonna roll up my sleeves this morning and dive in on this one.

File: BREEZE.rst
Comment (dimberman) [inline BREEZE.rst:58]: 🤦‍♂ *sigh*... I'll allow it.

File: BREEZE.rst
Comment (dimberman) [inline BREEZE.rst]: Do we still want to suppor 2.7 here?

**Review by dimberman — COMMENTED** [review]
First look-through this looks good, but I'm gonna hold off an "Approve" until I've had time to actually play with this environment.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Right :). Should be 3.7 in this branch :). I just enabled 3.7 master When I cherry-pick to v1-10-test it will be 2.7/3.5/3.6 there

**Comment by potiuk** [comment]
I simplified pylint - it is now fully integrated in pre-commit including filtering out files from pylint_todo list - that made it quite a bit simpler for running static checks with breeze (but now it is based on not-yet merged two commits with AIRFLOW-5285)

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Resolved

File: BREEZE.rst
Comment (bolkedebruin) [reply BREEZE.rst:58]: 😂😂😂😂😂

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
  you need to run ``brew install gnu-getopt`` and then follow instructions (you need
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]: Can we group this one with the 1st point about installing `gnu-getopt`

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst:123]: Awesome

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]: As they are commands:

```suggestion
The tests will run ``airflow db reset`` and ``airflow db init`` the first time you
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
Note that in case of ``nosetest`` you need to provide ``--nocapture`` flag to avoid nosetests capturing the stdout
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst:260]:
```suggestion
For databases you need to run ``airflow db reset`` at least once after you started Airflow Breeze to get
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]: Just curious why were are using raw html and not the `image` tag

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
When you run example DAGs, even if you run them using UnitTests from within IDE, they are run in a separate
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst:368]:
```suggestion
You can add arguments you should pass them after ``--`` as extra arguments.
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
``-x``, ``--execute-command`` flag. Note that if you want to add arguments you should specify them
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
together with the command surrounded with " or ' or pass them after -- as extra arguments.
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]: This is just my preference, not mandatory. You can ignore this, I don't have a strong opinion :)

```suggestion
``-d``, ``--docker-compose`` flag. Note that if you want to add extra arguments you should specify them
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst]:
```suggestion
These are the current flags of the `./breeze <./breeze>`_ script
```

File: BREEZE.rst
Comment (kaxil) [inline BREEZE.rst:564]: I don't think we should list all the commands in the `.rst` file. The main reason I think that is because if we change a flag in the original script or modify more flags it is easy to forget to update that in the doc.

And a user can simply get this by `breeze --help` and will probably make this file small & concise :)

**Review by kaxil — COMMENTED** [review]
Awesome job @potiuk . A big chunk of work, really appreciate it.

Small nits. I haven't yet tested it but the docs look really solid. Will add more comments once I do test it next week.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Done.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Done.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst:123]: And it also works for .zsh  (limited to options for now)

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Right. done

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst:260]: Done.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: It was automatically converted from an earlier markdonw version and I used raw there to have center alignment (no such feature in markdown). I have not noticed it used RAW :). Fixed now..

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Fixed.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Done

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Done.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: It makes perfect sense. I just missed it here, also have the same preference. Fixed.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Fixed.

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst:564]: It's generally nice if you can see all the commands in the web docs. I think this is something I always look for in this kind of documentation (there are people who prefer web docs than command line --help in general). Also it makes it google-searchable and web-viewable, which is rather nice if you are looking for some answers and you have no tools handy (like when you are answering someone's comments on slack and you only have your phone with you),

For example all gcloud commands have something like that: https://cloud.google.com/sdk/gcloud/reference/

I wanted to make it automated in the near future (like doctoc in .md files) so it does not get out-of-sync - this should be rather easy to do but it's already a huge change and I would love to get it out :).

File: BREEZE.rst
Comment (potiuk) [reply BREEZE.rst]: Resolved

**Comment by potiuk** [comment]
@kaxil @fokko @mik-laj @ashb I'd love to merge it today for the workshop today :). No huge problem if it is not merged but it would be great if I can show this from master.

**Review by mik-laj — APPROVED** [review]
Good luck!

**Review by mik-laj — APPROVED** [review]
Good luck!

**Comment by potiuk** [comment]
Woho! Finally !

**Pull request merged**
